addons/test_mail/tests/test_mail_followers_unsubscribe.py

```diff
@@ -46,3 +53,148 @@ def test_inbox_notification_follower(self):
         self.assertEqual(follower.res_id, test_record.id)
         self.assertEqual(follower.partner_id, self.partner_employee)
         test_record.with_user(self.user_admin).message_unsubscribe(partner_ids=self.partner_employee.ids)
+
+
+@tagged('mail_followers', 'post_install', '-at_install')
+class UnfollowFromEmailTest(TestMessagePostCommon, HttpCase):
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        cls.test_record = cls.env['mail.test.simple'].browse(cls.test_record.ids)
+        cls.test_record_admin = cls.test_record.with_user(cls.user_admin)
+        cls.test_record_unfollow = cls.env['mail.test.simple.unfollow'].create({'name': 'unfollow'})
+        cls.test_record_unfollow_admin = cls.test_record_unfollow.with_user(cls.user_admin)
+        cls.user_employee.write({'notification_type': 'email'})
+        cls.partner_admin = cls.user_admin.partner_id
+        cls.partner_without_user = cls.env['res.partner'].create({
+            'name': 'Dave',
+            'email': 'dave@odoo.com',
+        })
+
+    def test_initial_data(self):
+        """Test some initial value."""
+        self.assertTrue(self.user_employee._is_internal())
+        record_employee = self.test_record.with_user(self.user_employee)
+        record_employee.check_access_rights('read')
+        record_employee.check_access_rule('read')
+        mail_template_arch = self.env.ref('mail.mail_notification_layout').arch
+        self.assertIn('/mail/unfollow', mail_template_arch)
+        self.assertNotIn('/mail/unfollow', re.sub(_UNFOLLOW_REGEX, '', mail_template_arch))
+
+    def _extract_unfollow_url(self, mail_body):
+        urls = {link_url for _, link_url, _, _ in re.findall(tools.HTML_TAG_URL_REGEX, mail_body)
+                if '/mail/unfollow' in link_url}
+        n_url = len(urls)
+        self.assertLessEqual(n_url, 1)
+        return next(iter(urls), None)
+
+    def _message_post(self, record):
```

Comment:
Not a fan of this kind of helper, because with the same naming of "message_post", it actually returns a body of an implied first email ... if we really want it, maybe name it "_post_and_get_email_body", something like that ?

Reply:
Yes you’re right. I have merged it with extract_unfollow_url and named post_message_and_get_unfollow_urls and added a small docstring: “Post a message on the record for the partners and extract the unfollow URLs.”. So it reduces the number of helper function and make more obvious what they do.